File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/OSNotificationTracker.java
package com.onesignal.influence;

import android.support.annotation.NonNull;

import com.onesignal.OSLogger;
import com.onesignal.influence.model.OSInfluence;
import com.onesignal.influence.model.OSInfluenceChannel;
import com.onesignal.influence.model.OSInfluenceType;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

class OSNotificationTracker extends OSChannelTracker {

    public static final String TAG = OSNotificationTracker.class.getCanonicalName();
    private static final String DIRECT_TAG = "direct";
    private static final String NOTIFICATIONS_IDS = "notification_ids";
    private static final String NOTIFICATION_ID = "notification_id";

    OSNotificationTracker(@NonNull OSInfluenceDataRepository dataRepository, OSLogger logger) {
        super(dataRepository, logger);
    }

    @Override
    public String getIdTag() {
        return NOTIFICATION_ID;
    }

    @Override
    JSONArray getLastChannelObjectsReceivedByNewId(String id) {
        try {
            return getLastChannelObjects();
        } catch (JSONException exception) {
            logger.error("Generating Notification tracker getLastChannelObjects JSONObject ", exception);
            return new JSONArray();
        }
    }

    @Override
    JSONArray getLastChannelObjects() throws JSONException {
        return dataRepository.getLastNotificationsReceivedData();
    }

    @Override
    OSInfluenceChannel getChannelType() {
        return OSInfluenceChannel.NOTIFICATION;
    }

    @Override
    int getChannelLimit() {
        return dataRepository.getNotificationLimit();
    }

    @Override
    int getIndirectAttributionWindow() {
        return dataRepository.getNotificationIndirectAttributionWindow();
    }

    @Override
    void saveChannelObjects(JSONArray channelObjects) {
        dataRepository.saveNotifications(channelObjects);
    }

    @Override
    void initInfluencedTypeFromCache() {
        OSInfluenceType influenceType = dataRepository.getNotificationCachedInfluenceType();
        setInfluenceType(influenceType);

        if (influenceType.isIndirect())
            setIndirectIds(getLastReceivedIds());
        else if (influenceType.isDirect())
            setDirectId(dataRepository.getCachedNotificationOpenId());

        logger.debug("OneSignal NotificationTracker initInfluencedTypeFromCache: " + this.toString());
    }

    @Override
    void addSessionData(@NonNull JSONObject jsonObject, OSInfluence influence) {
        if (influence.getInfluenceType().isAttributed())
            try {
                jsonObject.put(DIRECT_TAG, influence.getInfluenceType().isDirect());
                jsonObject.put(NOTIFICATIONS_IDS, influence.getIds());
            } catch (JSONException exception) {
                logger.error("Generating notification tracker addSessionData JSONObject ", exception);
            }
    }

    @Override
    public void cacheState() {
        dataRepository.cacheNotificationInfluenceType(influenceType == null ? OSInfluenceType.UNATTRIBUTED : influenceType);
        dataRepository.cacheNotificationOpenId(directId);
    }

}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/data/OSNotificationTracker.kt
package com.onesignal.influence.data

import com.onesignal.OSLogger
import com.onesignal.OSTime
import com.onesignal.influence.OSInfluenceConstants
import com.onesignal.influence.domain.OSInfluence
import com.onesignal.influence.domain.OSInfluenceChannel
import com.onesignal.influence.domain.OSInfluenceType
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

internal class OSNotificationTracker(dataRepository: OSInfluenceDataRepository, logger: OSLogger, timeProvider: OSTime) : OSChannelTracker(dataRepository, logger, timeProvider) {
    override fun getLastChannelObjectsReceivedByNewId(id: String?): JSONArray {
        return try {
            lastChannelObjects
        } catch (exception: JSONException) {
            logger.error("Generating Notification tracker getLastChannelObjects JSONObject ", exception)
            JSONArray()
        }
    }

    @get:Throws(JSONException::class)
    override val lastChannelObjects: JSONArray
        get() = dataRepository.lastNotificationsReceivedData
    override val idTag: String
        get() = OSInfluenceConstants.NOTIFICATION_ID_TAG

    override val channelType: OSInfluenceChannel
        get() = OSInfluenceChannel.NOTIFICATION

    override val channelLimit: Int
        get() = dataRepository.notificationLimit

    override val indirectAttributionWindow: Int
        get() = dataRepository.notificationIndirectAttributionWindow

    override fun saveChannelObjects(channelObjects: JSONArray) {
        dataRepository.saveNotifications(channelObjects)
    }

    override fun initInfluencedTypeFromCache() {
        influenceType = dataRepository.notificationCachedInfluenceType.also {
            if (it.isIndirect())
                indirectIds = lastReceivedIds
            else if (it.isDirect())
                directId = dataRepository.cachedNotificationOpenId
        }
        logger.debug("OneSignal NotificationTracker initInfluencedTypeFromCache: $this")
    }

    override fun addSessionData(jsonObject: JSONObject, influence: OSInfluence) {
        if (influence.influenceType.isAttributed()) try {
            jsonObject.put(OSInfluenceConstants.DIRECT_TAG, influence.influenceType.isDirect())
            jsonObject.put(OSInfluenceConstants.NOTIFICATIONS_IDS, influence.ids)
        } catch (exception: JSONException) {
            logger.error("Generating notification tracker addSessionData JSONObject ", exception)
        }
    }

    override fun cacheState() {
        dataRepository.cacheNotificationInfluenceType(influenceType ?: OSInfluenceType.UNATTRIBUTED)
        dataRepository.cacheNotificationOpenId(directId)
    }
}